# MQB7046 MODELLING PUBLIC HEALTH DATA - Multinomial logistic regression

###### Prepared by Claire Choo (4/4/2024)

Multinomial logistic regression is a statistical method used to model the relationship between multiple categorical outcome variables and one or more independent variables.

* Data Preparation:
  - Load the dataset and ensure it is properly formatted.
  - Convert any categorical variables into dummy variables or use one-hot encoding.


* Model Specification:
  - Import the statsmodels.api module
  - Define independent variables and dependent variable
  - Add a constant termther independent variables using sm.add_constant().
  - Specify the multinomial logistic regression model using sm.MNLog().


* Model Fitting:
  - Fit the model with the data using the .fit() method.


* Interpretation and Evaluation: 
  - Interpret the coefficients, p-values, and confidence intervals to assess the significance of the independent variables (or predictors).
  - Assess the goodness-of-fit of the model using appropriate metrics like pseudo-R squared, likelihood ratio tests, or others.
  - Consider the odds ratios associated with each predictor variable to assess the magnitude of the effect.


#### Practical 4

The researchers are interested to examine factors associated with employment status of a group of individuals.

Variable / Definition:
1) id  : identification number
2) Sex : Gender of participants (Male, Female)
3) Age : Age group of participants (25-44,45-54, 55-64)
4) Marital: Marital status (Married/Cohabiting, Single, Widowed/Divorced) 
5) Employment : Employment status (Employed, Nonemployed, Unemployed)


In [3]:
import pandas as pd

# Load the dataset into a DataFrame

work = pd.read_csv("C:\\Users\\USER\\notebooks\\open.csv")

# Display the first few rows of the DataFrame to verify that the data is loaded correctly
print(work)

      id     Sex    Age             Marital   Employment
0      1    Male  55-64  Married/Cohabiting     Employed
1      2    Male  25-44  Married/Cohabiting  Nonemployed
2      3    Male  25-44  Married/Cohabiting     Employed
3      4    Male  55-64  Married/Cohabiting     Employed
4      5    Male  55-64  Married/Cohabiting     Employed
..   ...     ...    ...                 ...          ...
736  737    Male  45-54  Married/Cohabiting     Employed
737  738    Male  25-44              Single   Unemployed
738  739  Female  25-44              Single     Employed
739  740    Male  25-44  Married/Cohabiting     Employed
740  741    Male  45-54   Widowed/ Divorced     Employed

[741 rows x 5 columns]


In [2]:
# Check the structure of the dataset
print(work.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 741 entries, 0 to 740
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          741 non-null    int64 
 1   Sex         741 non-null    object
 2   Age         741 non-null    object
 3   Marital     741 non-null    object
 4   Employment  741 non-null    object
dtypes: int64(1), object(4)
memory usage: 29.1+ KB
None


In [3]:
# Convert X to dummy variables (#convert categorical variable(s) into dummy/indicator variables)
X_test1 = pd.get_dummies(work, columns=['Sex'], drop_first=True)

In [4]:
print(X_test1)

      id    Age             Marital   Employment  Sex_Male
0      1  55-64  Married/Cohabiting     Employed      True
1      2  25-44  Married/Cohabiting  Nonemployed      True
2      3  25-44  Married/Cohabiting     Employed      True
3      4  55-64  Married/Cohabiting     Employed      True
4      5  55-64  Married/Cohabiting     Employed      True
..   ...    ...                 ...          ...       ...
736  737  45-54  Married/Cohabiting     Employed      True
737  738  25-44              Single   Unemployed      True
738  739  25-44              Single     Employed     False
739  740  25-44  Married/Cohabiting     Employed      True
740  741  45-54   Widowed/ Divorced     Employed      True

[741 rows x 5 columns]


In [5]:
# selects the column named 'Sex_Male' from the DataFrame X_test1. It uses double square brackets [['Sex_Male']] 
# to specify that we are selecting a subset of columns and not just a single column.

X_Sex = X_test1[['Sex_Male']] 

In [6]:
# Convert the target variable to categorical type and then encode it to numerial codes according to the a specific order
y_test1 = pd.Categorical(work['Employment'], ordered=True, categories=['Employed', 'Nonemployed', 'Unemployed']).codes


In [7]:
print(y_test1)

[0 1 0 0 0 0 1 2 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0
 1 0 0 0 0 0 1 0 0 0 0 0 0 2 0 1 0 0 0 1 0 0 0 0 0 2 1 0 0 0 0 0 0 0 0 0 0
 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 2 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 1 0 0 1 0 1 0 0 0 1 0 0 0 0 1 0 1 2 1 1 0 0 0 0 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 2 0 0 1 0 0 0 0 1 0 0 1 0 1 1 1 1 0 0 1 1
 1 0 2 1 2 1 0 0 0 0 1 2 1 2 0 0 0 0 0 0 0 2 0 2 1 0 1 0 1 0 2 0 0 2 0 1 0
 0 0 0 0 0 0 1 0 1 0 1 0 0 1 1 1 0 1 2 0 0 0 1 2 0 0 2 0 0 1 1 0 1 0 1 0 0
 0 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 1 2 0 0 0 0 1 0 1 0 1 0 0 0 1 0 0 1
 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 2 2 1 0 0 0 1 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1
 0 1 1 0 0 0 0 0 0 1 1 0 1 0 2 0 0 0 0 1 1 2 0 2 0 1 0 0 0 0 0 0 0 1 0 0 0
 0 1 0 0 0 0 1 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0 0
 1 0 0 1 0 1 0 0 1 0 0 2 

#### Fitting multinomial logistic regression

In [8]:
# Fit multinomial logistic regression (1 IV and DV)

import statsmodels.api as sm

# Test 1: Single Independent Variable (Sex) with Single Target Variable
#add constant
X_test1 = sm.add_constant(X_Sex)


# Fit the multinomial logistic regression model
model_test1 = sm.MNLogit(y_test1, X_test1.astype(float))  # Ensure all variables are float
result_test1 = model_test1.fit()

# Print the summary of the model
print(result_test1.summary())

# Get AIC and BIC
aic = result_test1.aic
bic = result_test1.bic

print("AIC:", round(aic, 2))
print("BIC:", round(bic, 2))


Optimization terminated successfully.
         Current function value: 0.695885
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                      y   No. Observations:                  741
Model:                        MNLogit   Df Residuals:                      737
Method:                           MLE   Df Model:                            2
Date:                Thu, 04 Apr 2024   Pseudo R-squ.:                 0.02219
Time:                        21:23:42   Log-Likelihood:                -515.65
converged:                       True   LL-Null:                       -527.35
Covariance Type:            nonrobust   LLR p-value:                 8.291e-06
       y=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9200      0.114     -8.085      0.000      -1.143      -0.697
Sex_Male      -0.8681      0.

##### Ouput

Coefficients:
- The 'Sex_Male' row represents the coefficient for the predictor variable 'Sex_Male', indicating the effect of being male compared to being female on the log-odds of belonging to each category of 'y'.
  
- For the category 'y=1' (non-employed): The negative coefficient (-0.8681) for 'Sex_Male' indicates that being male (compared to being female, the reference category) decreases the log-odds of being non-employed (y=1). In other words, being male is associated with a lower likelihood of being non-employed compared to being female.
  
- For the category 'y=2' (unemployed): The positive coefficient (0.3302) for 'Sex_Male' suggests that being male (compared to being female, the reference category) increases the log-odds of being unemployed (y=2). However, this effect is not statistically significant at conventional significance levels (p > 0.05), meaning that there is insufficient evidence to conclude that gender significantly influences the likelihood of being unemployed in this model.

Model comparison: 
- `AIC (Akaike Information Criterion)`: A measure of model fit that penalizes for model complexity. Lower values indicate better fi 
- `BIC (Bayesian Information Criterion)`: Similar to AIC but with a stronger penalty for model complexity. Again, lower values indicate better fit
- `LLR p-value`: Indicates the statistical significance of the model compared to a model with no predictors (null model). In this case, the LLR p-value is very small, indicating that the model significantly improves the fit compared to the null model.
 

In [9]:
import numpy as np

# Coefficients for y=1 (Nonemployed)
const_coef_y1 = -0.9200
sex_male_coef_y1 = -0.8681

# Coefficients for y=2 (Unemployed)
const_coef_y2 = -2.7689
sex_male_coef_y2 = 0.3302

# Compute odds ratios
odds_ratio_y1 = np.exp(sex_male_coef_y1)
odds_ratio_y2 = np.exp(sex_male_coef_y2)

print("Odds Ratio for Nonemployed(Sex_Male):", round(odds_ratio_y1, 2))
print("Odds Ratio for Unemployed (Sex_Male):", round(odds_ratio_y2, 2))


Odds Ratio for Nonemployed(Sex_Male): 0.42
Odds Ratio for Unemployed (Sex_Male): 1.39


In [10]:
import numpy as np

# Coefficients for y=1 (Nonemployed)
const_coef_y1 = -0.9200
sex_male_coef_y1 = -0.8681
sex_male_std_err_y1 = 0.196  # Standard error for Sex_Male coefficient in y=1

# Coefficients for y=2 (unemployed)
const_coef_y2 = -2.7689
sex_male_coef_y2 = 0.3302
sex_male_std_err_y2 = 0.328  # Standard error for Sex_Male coefficient in y=2

# Compute odds ratios
odds_ratio_y1 = np.exp(sex_male_coef_y1)
odds_ratio_y2 = np.exp(sex_male_coef_y2)

# Compute 95% confidence intervals
ci_low_y1 = np.exp(sex_male_coef_y1 - 1.96 * sex_male_std_err_y1)
ci_high_y1 = np.exp(sex_male_coef_y1 + 1.96 * sex_male_std_err_y1)

ci_low_y2 = np.exp(sex_male_coef_y2 - 1.96 * sex_male_std_err_y2)
ci_high_y2 = np.exp(sex_male_coef_y2 + 1.96 * sex_male_std_err_y2)

print("Odds Ratio for y=1 (Sex_Male):", round(odds_ratio_y1, 2))
print("95% CI for y=1 (Sex_Male): [{:.2f}, {:.2f}]".format(ci_low_y1, ci_high_y1))

print("Odds Ratio for y=2 (Sex_Male):", round(odds_ratio_y2, 2))
print("95% CI for y=2 (Sex_Male): [{:.2f}, {:.2f}]".format(ci_low_y2, ci_high_y2))



Odds Ratio for y=1 (Sex_Male): 0.42
95% CI for y=1 (Sex_Male): [0.29, 0.62]
Odds Ratio for y=2 (Sex_Male): 1.39
95% CI for y=2 (Sex_Male): [0.73, 2.65]


##### Output:

For the category 'y=1' (non-employed):

Odds Ratio: The odds ratio for 'Sex_Male' in category 'y=1' is 0.42. The odds of being non-employed (y=1) for males are 0.42 times compared to females. Being male is associated with 58% lower likelihood of being non-employed compared to being female.
95% Confidence Interval (CI): 95% CI for the odds ratio of 'Sex_Male' in category 'y=1' is [0.29, 0.62]. Since the interval does not include 1, suggests that the effect of gender on non-employment is statistically significant.

For the category 'y=2' (unemployed):
Odds Ratio: The odds ratio for 'Sex_Male' in category 'y=2' is 1.39. The odds of being unemployed (y=2) for males are 1.39 times the odds for females, holding other variables constant. Being male is associated with 39% higher likelihood of being unemployed compared to being female.
95% Confidence Interval (CI): 95% CI for the odds ratio of 'Sex_Male' in category 'y=2' is [0.73, 2.65]. Since the interval includes 1, suggests that the effect of gender on unemployment is not statistically significant at conventional significance levels.

In [11]:
#Changing reference group of Sex to Male
# Convert X to dummy variables
X_test2 = pd.get_dummies(work, columns=['Sex'], drop_first=False)
X_Sexf = X_test2[['Sex_Female']]
print(X_Sexf)

     Sex_Female
0         False
1         False
2         False
3         False
4         False
..          ...
736       False
737       False
738        True
739       False
740       False

[741 rows x 1 columns]


In [12]:
import statsmodels.api as sm

# Test 1: Single Independent Variable with Single Target Variable
#add constant
X_test2 = sm.add_constant(X_Sexf)

# Fit the multinomial logistic regression model
model_test2 = sm.MNLogit(y_test1, X_test2.astype(float))  # Ensure all variables are float
result_test2 = model_test2.fit()

# Print the summary of the model
print(result_test2.summary())

# Get AIC and BIC
aic = result_test2.aic
bic = result_test2.bic

print("AIC:", round(aic, 2))
print("BIC:", round(bic, 2))


Optimization terminated successfully.
         Current function value: 0.695885
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                      y   No. Observations:                  741
Model:                        MNLogit   Df Residuals:                      737
Method:                           MLE   Df Model:                            2
Date:                Thu, 04 Apr 2024   Pseudo R-squ.:                 0.02219
Time:                        21:23:42   Log-Likelihood:                -515.65
converged:                       True   LL-Null:                       -527.35
Covariance Type:            nonrobust   LLR p-value:                 8.291e-06
       y=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.7881      0.159    -11.225      0.000      -2.100      -1.476
Sex_Female     0.8681      0.

In [13]:
### Adding more IVs in the model

In [14]:
# Convert selected columns to dummy variables
columns_to_convert = ['Sex', 'Marital', 'Age']
X_test3 = pd.get_dummies(work, columns=columns_to_convert, drop_first=True)
print(X_test3)

      id   Employment  Sex_Male  Marital_Single  Marital_Widowed/ Divorced  \
0      1     Employed      True           False                      False   
1      2  Nonemployed      True           False                      False   
2      3     Employed      True           False                      False   
3      4     Employed      True           False                      False   
4      5     Employed      True           False                      False   
..   ...          ...       ...             ...                        ...   
736  737     Employed      True           False                      False   
737  738   Unemployed      True            True                      False   
738  739     Employed     False            True                      False   
739  740     Employed      True           False                      False   
740  741     Employed      True           False                       True   

     Age_45-54  Age_55-64  
0        False       True  
1      

In [15]:
# Drop 'id' and 'Employment' columns from X_test1
X_test3 = X_test3.drop(['id', 'Employment'], axis=1)
print(X_test3)

     Sex_Male  Marital_Single  Marital_Widowed/ Divorced  Age_45-54  Age_55-64
0        True           False                      False      False       True
1        True           False                      False      False      False
2        True           False                      False      False      False
3        True           False                      False      False       True
4        True           False                      False      False       True
..        ...             ...                        ...        ...        ...
736      True           False                      False       True      False
737      True            True                      False      False      False
738     False            True                      False      False      False
739      True           False                      False      False      False
740      True           False                       True       True      False

[741 rows x 5 columns]


In [16]:
import statsmodels.api as sm

# Test 1: Single Independent Variable with Single Target Variable
#add constant
X_test3 = sm.add_constant(X_test3)


# Fit the multinomial logistic regression model
model_test3 = sm.MNLogit(y_test1, X_test3.astype(float))  # Ensure all variables are float
result_test3 = model_test3.fit()


# Get the summary of the model
result_summary = result_test3.summary()


# Print the summary of the model
print(result_test3.summary())

# Get AIC and BIC
aic = result_test3.aic
bic = result_test3.bic

print("AIC:", round(aic, 2))
print("BIC:", round(bic, 2))


Optimization terminated successfully.
         Current function value: 0.646842
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                      y   No. Observations:                  741
Model:                        MNLogit   Df Residuals:                      729
Method:                           MLE   Df Model:                           10
Date:                Thu, 04 Apr 2024   Pseudo R-squ.:                 0.09110
Time:                        21:23:42   Log-Likelihood:                -479.31
converged:                       True   LL-Null:                       -527.35
Covariance Type:            nonrobust   LLR p-value:                 3.305e-16
                      y=1       coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
const                        -1.2499      0.170     -7.357      0.000      -1.

Example of interpretation for exposure/independent variables that has more than two categories (e.g, age group):

Interpretation for 'y=1' (non-employed):
For the 'Age_45-54' group, the coefficient is -0.3805. Individuals aged 45-54 have lower log-odds of being non-employed compared to the reference group (youngest age group, 44 and below), given other variables constant. This effect is not statistically significant p > 0.05.
For the 'Age_55-64' group, the coefficient is 1.1645. Individuals aged 55-64 have higher log-odds of being non-employed compared to the reference group, and this effect is statistically significant (p < 0.001).

Interpretation for 'y=2' (unemployed):
For both the 'Age_45-54' and 'Age_55-64' groups, the coefficients are negative (-0.2907 and -0.2116, respectively), indicating that individuals in these age groups have lower log-odds of being unemployed compared to the reference group. However, these effects are not statistically significant at conventional significance levels (p > 0.05).



In [17]:
# Comparing model 2 and model 1 using AIC, BIC

# Import the necessary libraries
from statsmodels.stats.anova import anova_lm

# Calculate the differences
delta_aic = result_test3.aic - result_test1.aic
delta_bic = result_test3.bic - result_test1.bic

print("Change in AIC:", round(delta_aic, 2))
print("Change in BIC:", round(delta_bic, 2))


Change in AIC: -56.68
Change in BIC: -19.82


In [18]:
# Import the necessary libraries

from scipy.stats import chi2

# Assume result_test1 is the previous model and result_test3 is the current model

# Compute the log-likelihoods for the two models
ll_reduced = result_test1.llf
ll_full = result_test3.llf

# Compute the likelihood ratio test statistic
lrt_statistic = -2 * (ll_reduced - ll_full)

# Degrees of freedom is equal to the difference in the number of parameters between the two models
df = result_test3.df_model - result_test1.df_model

# Then compare lrt_statistic to the chi-square distribution with df degrees of freedom to get the p-value
# Calculate the p-value *cdf - Cumulative Distribution Function.
p_value = 1 - chi2.cdf(lrt_statistic, df)

# Print the results
print("Likelihood Ratio Test Results")
print("Test statistic:", round(lrt_statistic, 2))
print("Degrees of Freedom:", df)
print("P-value:", p_value)

Likelihood Ratio Test Results
Test statistic: 72.68
Degrees of Freedom: 8.0
P-value: 1.4345191701181648e-12


Since the p-value is very small (< 0.05), we reject the null hypothesis.
The full model provides a significantly better fit to the data compared to the reduced model.
Including the additional variables in the full model improves its ability to explain the variability in the response variable compared to the reduced model.